### PR TITLE
fix: Expensify card link navigates to wallet

### PR DESCRIPTION
**Pull request by truph01**
### Fixed Issues

$ https://github.com/Expensify/App/issues/49804
PROPOSAL: https://github.com/Expensify/App/issues/49804#issuecomment-2378111577

### Tests

1. Go to ND and log in
2. Create a workspace
3. Navigate to the workspace settings > More Features
4. Enable Expensify card option
5. Navigate to the Expensify card and click on Issue card
6. Select Connect online with plaid and add the VBA
7. Issue the virtual Expensify card to the current user
8. Navigate to the workspace chat
9. Click on the Expensify card link
10. Verify: User is navigated to the virtual Expensify card details
- [x] Verify that no errors appear in the JS console

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/dee5f1b8-46d8-4d57-8b3c-9d01d63cde01

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/d7e14f14-e343-46df-9774-6b45004e7383

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/a2561444-88b3-4cb4-8ea9-7ca0ce1c1c09

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/73521fae-ca81-4ddf-bcd1-6e4d701a7cf1

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/26245782-1df7-4099-84ef-add3c5bb43b3

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/bfee2989-4165-439a-9cae-4f7eec9a245b

</details>

### CODE REVIEW

**Comment by truph01** [comment]
I cannot run android native. Will update the video later.

**Comment by jjcoffee** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by jjcoffee** [comment]
@truph01 Were you able to actually test this? I can't seem to get a card added.

**Comment by jjcoffee** [comment]
`BUG:` If the user the card is issued to is not the admin, clicking the link from the workspace chat (as that non-admin user) they are navigated to a `Not found` page. The route in this case should be `settings/wallet/card/${cardID}`.

https://github.com/user-attachments/assets/be20d308-19d1-4a21-8618-a8a9b29c3755

**Comment by jjcoffee** [comment]
`BUG:` On first sign-in, clicking the link loads the card details but the cards page behind the modal is blank and has a jittery render when dismissing the modal. This is because normally we'd call `OpenPolicyExpensifyCardsPage` first when this page is opened "normally".

https://github.com/user-attachments/assets/644c758c-2c63-4a30-9a86-a46246a44175

**Review by jjcoffee — CHANGES_REQUESTED** [review]
A couple of bugs to sort out!

**Comment by truph01** [comment]
> BUG: If the user the card is issued to is not the admin, clicking the link from the workspace chat (as that non-admin user) they are navigated to a Not found page. The route in this case should be settings/wallet/card/${cardID}.

Fixed this bug.

- Member:

https://github.com/user-attachments/assets/9c18f8c9-0121-4e8a-b3fc-5d4e7836b7c4

- Admin:

https://github.com/user-attachments/assets/19a0de17-920d-4350-959f-9c9d0f3a4576

**Comment by truph01** [comment]
> BUG: On first sign-in, clicking the link loads the card details but the cards page behind the modal is blank and has a jittery render when dismissing the modal. This is because normally we'd call OpenPolicyExpensifyCardsPage first when this page is opened "normally".

When we are on the report page and then click on the card link, it will navigate to the card detail page in RHN. But in background, which screen should be displayed? There are two options: 

1. Currently, in PR, it is:



https://github.com/user-attachments/assets/ae767500-b84d-4c88-9af5-89c87340a5e0

2. In the background, the report is displayed.


@iwiznia @jjcoffee

**Comment by jjcoffee** [comment]
> When we are on the report page and then click on the card link, it will navigate to the card detail page in RHN. But in background, which screen should be displayed?

I think it probably makes sense to have the report in the background (central pane), unless it adds unnecessary complexity. Let's do that for now, we can revert if @iwiznia feels strongly about it!

**Comment by jjcoffee** [comment]
@truph01 Could you also add the Android native video when you get a chance? Thanks!

**Comment by iwiznia** [comment]
It's probably a bit better to keep the report (or whatever page you were in) in the background, so when you close the RHP you are in the same place you were before

**Comment by truph01** [comment]
I cannot receive magic code now. Did you encounter the same? @jjcoffee

**Comment by jjcoffee** [comment]
@truph01 Should be fixed now!

**Comment by truph01** [comment]
> It's probably a bit better to keep the report (or whatever page you were in) in the background, so when you close the RHP you are in the same place you were before

- Updated. Below is result:

#### Admin:


https://github.com/user-attachments/assets/7ccf4232-1515-495c-b79b-68e9c04f0156


#### Others:


https://github.com/user-attachments/assets/65a2b47a-8c67-4b3a-ae45-981e1395c6da

**Comment by truph01** [comment]
@iwiznia Also, please help confirm the @jjcoffee's thoughts:

> BUG: If the user the card is issued to is not the admin, clicking the link from the workspace chat (as that non-admin user) they are navigated to a Not found page. The route in this case should be settings/wallet/card/${cardID}.

**Comment by truph01** [comment]
> @truph01 Could you also add the Android native video when you get a chance? Thanks!



@jjcoffee I updated android video

File: src/libs/ReportActionsUtils.ts
Comment (jjcoffee) [inline src/libs/ReportActionsUtils.ts:1734]: Can we avoid casting here?

**Comment by jjcoffee** [comment]
@truph01 Testing on Android native, when I tap the link as an admin (the card is assigned to someone else), the not found page shows up. This also happens for the user the card is assigned to. The route in the logs:

image

**Comment by truph01** [comment]
> Testing on Android native, when I tap the link as an admin (the card is assigned to someone else), the not found page shows up. This also happens for the user the card is assigned to. The route in the logs:

I encountered the same when I tested on Android. I try to call "npm run android" and start metro "npx react-native start :8081" again, then it works properly. Can you try it?

File: src/libs/ReportActionsUtils.ts
Comment (truph01) [reply src/libs/ReportActionsUtils.ts]: I fixed

File: src/libs/ReportActionsUtils.ts
Comment (jjcoffee) [inline src/libs/ReportActionsUtils.ts:1733]: Let's DRY this up and put the `isActionOfType` stuff into a const.

**Review by jjcoffee — APPROVED** [review]
LGTM!

File: src/libs/ReportActionsUtils.ts
Comment (iwiznia) [inline src/libs/ReportActionsUtils.ts:1732]: Why -1 and not 0 or null? (same in all the ones below)

File: src/libs/ReportActionsUtils.ts
Comment (iwiznia) [inline src/libs/ReportActionsUtils.ts:1745]: I think this is wrong if policyID is -1, no?

File: src/libs/ReportActionsUtils.ts
Comment (truph01) [reply src/libs/ReportActionsUtils.ts:1732]: Because the ```policyID``` prop of the ```report``` is ```policyID?: string```, so I need to add the fallback value as '-1' to fix the type error based on the docs: https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#default-value-for-inexistent-ids

File: src/libs/ReportActionsUtils.ts
Comment (truph01) [reply src/libs/ReportActionsUtils.ts:1745]: Yep, ```isPolicyAdmin``` will be ```false``` if policyID is -1

Comment (iwiznia) [reply src/libs/ReportActionsUtils.ts:1745]: This sounds wrong, because `-1` is not checked here https://github.com/Expensify/App/blob/9e1f21ac6c354988b45edfa88eb0950b240949c7/src/libs/PolicyUtils.ts#L579 😕 
But also checking with the team about the -1 reocommendation

Comment (truph01) [reply src/libs/ReportActionsUtils.ts:1745]: > This sounds wrong, because -1 is not checked here


- If policyID = -1, then condition:

https://github.com/Expensify/App/blob/9e1f21ac6c354988b45edfa88eb0950b240949c7/src/libs/PolicyUtils.ts#L579
is false

so undefined will be returned because of:

https://github.com/Expensify/App/blob/9e1f21ac6c354988b45edfa88eb0950b240949c7/src/libs/PolicyUtils.ts#L582

I think ```getPolicy(-1) = undefined``` makes sense.

> But also checking with the team about the -1 recommendation you linked above


- Waiting for the final decision

File: src/libs/ReportActionsUtils.ts
Comment (iwiznia) [reply src/libs/ReportActionsUtils.ts:1732]: Posted about this here https://expensify.slack.com/archives/C01GTK53T8Q/p1728329641957179 and we will be undoing this, but I'll go ahead and merge this

**Review by iwiznia — APPROVED** [review]

**Comment by jjcoffee** [comment]
Coming from https://github.com/Expensify/App/issues/50375, we missed handling cases where the card is no longer active/available. More details [here](https://github.com/Expensify/App/issues/50375#issuecomment-2399546386).

**Pull request merged**
